packages/gatsby/src/redux/actions/public.js

```diff
@@ -47,6 +48,19 @@ const ensureWindowsDriveIsUppercase = filePath => {
     : filePath
 }
 
+const truncatePath = path => {
+  if (path === `/`) return path
+  const pathParts = path.split(`/`).filter(Boolean)
+  const truncatedParts = pathParts.map(pathPart => {
+    // OS limit is 255 characters
+    if (pathPart.length > 255) {
+      return pathPart.slice(0, 100) + strhash(pathPart.slice(100))
+    }
+    return pathPart
+  })
+  return truncatedParts.join(`/`)
```

Comment:
we should return exact input if no segment was changed - this make it a different path even if hash wasn't used in this path

the slashes are tricky - this essentially will remove trailing slashes if the `truncateLongPaths` is enabled for all paths

How about using regex here to target all segments (segment would be consecutive characters that don't include slashes). Example below also using `createContentDigest` instead of additional dependency


```js
const { createContentDigest } = require(`gatsby-core-utils`)

const MAX_PATH_SEGMENT = 255
const SLICING_INEDX = 100
const pathSegmentRe = /[^\/]+/g

const truncatePath = path => {
  return path.replace(pathSegmentRe, match => {
    if (match.length > MAX_PATH_SEGMENT) {
      return match.slice(0, SLICING_INEDX) + createContentDigest(match.slice(SLICING_INEDX))
    }
    
    return match
  })
}
```

Check https://runkit.com/pieh/5e526ed73e251d001bdd7e19 how that would behave for various inputs (with shorter cutoff constants for testing purposes